src/v/utils/functional.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "utils/concepts-enabled.h"
+
+#include <functional>
+#include <optional>
+namespace reduce {
+struct push_back {
+    /// TODO(rob) introduce a c++ concept here
+    template<typename vec_like>
+    vec_like operator()(vec_like acc, typename vec_like::value_type t) {
+        acc.push_back(std::move(t));
+        return std::move(acc);
+    }
+};
+
+struct push_back_opt {
```

Comment:
push all of these struct helpers to `namespace reduce::details;`

Reply:
@senior7515 why did you suggest they go in a details namespace? Consumers that include this source file are expected to use these utils, they aren't scoped locally to this source file only